tests/test_providers.py

```diff
@@ -263,3 +275,464 @@ def parse_(self, response: DummyResponse, page: AnnotatedProductPage):  # type:
+    assert params[0].keys() == {"url", "product", "browserHtml", "productOptions"}
+
+    assert type(item["page"].response) == AnyResponse
+    assert type(item["page"].response.response) == BrowserResponse
+    assert type(item["page"].product) == Product
+
+
+@ensureDeferred
+async def test_provider_any_response_product_extract_from_browser_html_2(mockserver):
+    @attrs.define
+    class SomePage(BasePage):
+        response: AnyResponse
+        browser_response: BrowserResponse
+        product: Product
+
+    class ZyteAPISpider(Spider):
+        def start_requests(self):
+            yield Request(self.url, callback=self.parse_)
+
+        def parse_(self, response: DummyResponse, page: SomePage):
+            yield {"page": page}
+
+    settings = provider_settings(mockserver)
+    settings["ZYTE_API_PROVIDER_PARAMS"] = {
+        "productOptions": {"extractFrom": "browserHtml"}
+    }
+    item, url, crawler = await crawl_single_item(ZyteAPISpider, HtmlResource, settings)
+    params = crawler.engine.downloader.handlers._handlers["http"].params
+
+    assert len(params) == 1
+    assert params[0].keys() == {"url", "product", "browserHtml", "productOptions"}
+
+    assert type(item["page"].response) == AnyResponse
+    assert type(item["page"].response.response) == BrowserResponse
+    assert type(item["page"].browser_response) == BrowserResponse
+    assert type(item["page"].product) == Product
+
+    assert id(item["page"].browser_response) == id(item["page"].response.response)
+
+
+@ensureDeferred
+async def test_provider_any_response_product_extract_from_http_response(mockserver):
+    @attrs.define
+    class SomePage(BasePage):
+        response: AnyResponse
+        product: Product
+
+    class ZyteAPISpider(Spider):
+        def start_requests(self):
+            yield Request(self.url, callback=self.parse_)
+
+        def parse_(self, response: DummyResponse, page: SomePage):
+            yield {"page": page}
+
+    settings = provider_settings(mockserver)
+    settings["ZYTE_API_PROVIDER_PARAMS"] = {
+        "productOptions": {"extractFrom": "httpResponseBody"}
+    }
+    item, url, crawler = await crawl_single_item(ZyteAPISpider, HtmlResource, settings)
+    params = crawler.engine.downloader.handlers._handlers["http"].params
+
+    assert len(params) == 1
+    assert params[0].keys() == {
+        "url",
+        "product",
+        "httpResponseBody",
+        "productOptions",
+        "httpResponseHeaders",
+        "customHttpRequestHeaders",
+    }
+
+    assert type(item["page"].response) == AnyResponse
+    assert type(item["page"].response.response) == HttpResponse
+    assert type(item["page"].product) == Product
+
+
+# The issue here is that HttpResponseProvider runs earlier than ScrapyZyteAPI.
+# HttpResponseProvider doesn't know that it should not run since ScrapyZyteAPI
+# could provide HttpResponse in anycase.
+@pytest.mark.xfail(reason="Not supported yet", raises=AssertionError, strict=True)
```

Comment:
This is something that might need more changes and can be handled in another PR.

Reply:
Should we document the disabling of the HttpResponseProvider as part of scrapy-zyte-api configuration?

Reply:
No, as that would prevent certain cases from being supported. For example: https://github.com/scrapy-plugins/scrapy-zyte-api/pull/161/files/932262288789a774476ca3d6474caf5d925703dc#r1464355664.

Could also be the case that `HttpResponse` comes from another source (e.g. Crawlera, though least likely).